ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Genesis/Governor.hs

```diff
@@ -100,3 +126,200 @@ reprocessLoEBlocksOnCandidateChange chainDb getCandidates =
+            \
+    frag2:     E           claimed tip: E <- exact
+
+-}
+
+data TraceGDDEvent peer blk =
+  TraceGDDEvent {
+    bounds            :: Map peer (DensityBounds blk),
+    candidateSuffixes :: Map peer (AnchoredFragment (Header blk)),
+    losingPeers       :: [peer],
+    loeHead           :: AF.Anchor (Header blk),
+    sgen              :: GenesisWindow
+  }
+
+-- | Run the Genesis disconnection logic once.
+--
+--   * Update the LoE and return it.
+--
+--   * Disconnect from peers with inferior density.
+updateLoEFragGenesis ::
+     forall m blk peer.
+     ( IOLike m
+     , Ord peer
+     , LedgerSupportsProtocol blk
+     , HasHardForkHistory blk
+     )
+  => TopLevelConfig blk
+  -> Tracer m (TraceGDDEvent peer blk)
+  -> STM m (Map peer (AnchoredFragment (Header blk)))
+  -> STM m (Map peer (ChainSyncClientHandle m blk))
+  -> UpdateLoEFrag m blk
+updateLoEFragGenesis cfg tracer getCandidates getHandles =
+  UpdateLoEFrag $ \ curChain immutableLedgerSt -> do
+    (candidateSuffixes, handles, theirTips, latestSlots, loeFrag) <- atomically $ do
+      candidates <- getCandidates
+      handles <- getHandles
+      let
+        (loeFrag, candidateSuffixes) =
+          sharedCandidatePrefix curChain candidates
+      theirTips <-
+        fmap (Map.mapMaybe id) $ flip Map.traverseWithKey candidateSuffixes $ \peer _ ->
+          cschTheirTip (handles Map.! peer)
+      latestSlots <-
+        flip Map.traverseWithKey candidateSuffixes $ \peer _ ->
+          cschLatestSlot (handles Map.! peer)
+      pure (candidateSuffixes, handles, theirTips, latestSlots, loeFrag)
+
+    let msgen :: Maybe GenesisWindow
+        -- This could also use 'runWithCachedSummary' if deemed desirable.
+        msgen = eitherToMaybe $ runQuery qry summary
+          where
+            -- We use the Genesis window for the first slot /after/ the common
+            -- intersection. In particular, when the intersection is the last
+            -- slot of an era, we will use the Genesis window of the next era,
+            -- as all slots in the Genesis window reside in that next era.
+            slot    = succWithOrigin $ AF.headSlot loeFrag
+            qry     = qryFromExpr $ slotToGenesisWindow slot
+            summary =
+              hardForkSummary
+                (configLedger cfg)
+                -- Due to the cross-chain lemma (Property 17.3 in the Consensus
+                -- report) one could also use the ledger state at the tip of our
+                -- selection here (in which case this should never return
+                -- 'Nothing'), but this is subtle and maybe not desirable.
+                --
+                -- In any case, the immutable ledger state will also
+                -- /eventually/ catch up to the LoE tip, so @msgen@ won't be
+                -- 'Nothing' forever.
+                (ledgerState immutableLedgerSt)
+
+    whenJust msgen $ \sgen -> do
+      let
+        (losingPeers, bounds) =
```

Comment:
How about we run a fixpoint computation here, so we recompute `loeFrag` and `losingPeers` until no more peers are disconnected?